lib/execution_segment.go

```diff
@@ -384,13 +385,82 @@ func (ess ExecutionSegmentSequence) String() string {
 	return strings.Join(result, ",")
 }
 
+func (ess ExecutionSegmentSequence) Len() int {
+	return len(ess)
+}
+
+func (ess ExecutionSegmentSequence) Less(i, j int) bool {
+	// Yes this Less is actually More, but we want it sorted in descending order and the alternative is to reverse it after sort
+	return ess[i].length.Cmp(ess[j].length) > 0
+}
+
+func (ess ExecutionSegmentSequence) Swap(i, j int) {
+	ess[i], ess[j] = ess[j], ess[i]
+}
+
+var _ sort.Interface = ExecutionSegmentSequence{}
+
+// lowest common denominator based on https://rosettacode.org/wiki/Least_common_multiple#Go
+func (ess ExecutionSegmentSequence) lcd() int64 {
+	var m, n, z big.Int
+	z = *ess[0].length.Denom()
+	for _, seg := range ess[1:] {
+		m = z
+		n = *seg.length.Denom()
+		z.Mul(z.Div(&m, z.GCD(nil, nil, &m, &n)), &n)
+	}
+
+	return z.Int64()
+}
+
 // GetStripedOffsets returns everything that you need in order to execute only
 // the iterations that belong to the supplied segment...
 //
 // TODO: add a more detailed algorithm description
-func (ess ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int, []int, error) {
-	start := 0
-	offsets := []int{}
-	// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
-	return start, offsets, fmt.Errorf("not implemented")
+// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
+func (ess ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int64, []int64, error) {
+	// TODO check if ExecutionSegmentSequence actually starts from 0 and goes to 1 ?
```

Comment:
:+1: this is not always going to be the case. While I was creating the algorithm, my thinking was that we can just pad the sequence with as many `1/lcd()` segments as needed, to complete their sum to `1`.

Reply:
Hm ... I really don't want to make this function more complicated ... Especially if we decide to optimize it even more (I think the sorting can be done based on the numerators but I will need to at the same time sort them and the other slices ... which means I will need to implement [sort.Interface](https://golang.org/pkg/sort/#Interface) on top of something ... so probably later). But I digress ... 

Can we agree that there are only two cases for `ess`:
1. `ess` is actually from 0 to 1  - currently handled by the code
2. `len(ess) == 0` (whether it's nil or just an empty slice is not relevant) 

In case 2 I will get the Denominator set the LCD to it and add as much to the numerators as needed, without actually populating `ess`. 
This is predominantly because in the case of the segment `1/6:3/6` ( this is `1/3` in reality I think it should be the same as `0:1/3` and `2/3:1`)  in order to populate ess correctly, I will need to add `0:1/6` and then `3/6:4/6`, `4/6:5/6` and `5/6,1` ... which IMO will be harder ... So "simple" padding won't work, and for pathological cases, this will be awful ... I don't think someone will use it like that but I think it should work in an easy to explain way :). 

I don't think that half-filled `ess` is a case that we will have? Like what is `1/6:3/6,3/6:4/6` with segment `3/6:4/6` suppose to mean?

Reply:
My idea implemented in 12246b4 , although now this defenitely needs refactoring ...

Reply:
:+1: :+1: for the refactoring :sweat_smile: 

I don't fully grok the current code, but I think we have 4 distinct scenarios that we need to tackle here:
1) No `ExecutionSegment` (and no `ExecutionSegmentSequence`) - this is handled by the executors, just means that there's no striping of any kind, so we don't have to concern ourselves with it in `ExecutionSegmentSequence`
2) Specified `ExecutionSegment` and **no** `ExecutionSegmentSequence` - this can be handled by a very fast algorithm that doesn't require the complexity of the current one. Say, if you have `ExecutionSegment=(0:1/3]`, then you just execute every 3rd iteration :sweat_smile: Or, maybe even easier, just multiply the arrival-rate executor's `Rate/TimeUnit` by `1/3`. No need to deal with or pad `ExecutionSegmentSequence` in any way.
3) Specified `ExecutionSegment` and an `ExecutionSegmentSequence` that spans the whole interval from `0` to `1`. This was the old algorithm, no need to pad anything with `1/lcd` pieces.
4) Specified `ExecutionSegment` and an `ExecutionSegmentSequence` that **doesn't** span the whole interval from `0` to `1`. For example, `ExecutionSegment=1/4:1/2` and `ExecutionSegmentSequence=1/4,1/2,1`. Then you need to pad the sequence with the missing pieces, `0:1/4` in this case. But for better code, you can do that outside of the `GetStripedOffsets()` function, probably...

Reply:
Okay so:
1. is fine :)
2. is currently handled in 12246b4 by (as I explained previously) [not padding ExecutionSegmentSequence but padding everything else that is populated from it(the numerators and steps slices)](https://github.com/loadimpact/k6/pull/1323/files#diff-4d3a948d3480dfb221e624ec54a23a34R446-R459)
3. this was basically handled previously as well and is now mostly [this code](https://github.com/loadimpact/k6/pull/1323/files#diff-4d3a948d3480dfb221e624ec54a23a34R418-R444)
4. I don't think this is a use case we have ... like what is the idea here? Either way, can we leave it for another PR later on if it's going to be done in a separate function outside of `GetStripedOffsets`?

So if you don't want to have 2. [handled in GetStrippedOffsets code](https://github.com/loadimpact/k6/pull/1323/files#diff-4d3a948d3480dfb221e624ec54a23a34R446-R459) I can just revert 12246b4 which will significantly lower the amount of code (as it will remove the whole else clause :)) 

BTW for easier understanding of what is happening, I propose going commit by commit as most changes are optimizations and are kind of hard to understand with all the code around :)

Reply:
> 2. is currently handled in [12246b4](https://github.com/loadimpact/k6/commit/12246b4449fbf22b295928a476aecf1771a10fdc) by (as I explained previously) [not padding ExecutionSegmentSequence but padding everything else that is populated from it(the numerators and steps slices)](https://github.com/loadimpact/k6/pull/1323/files#diff-4d3a948d3480dfb221e624ec54a23a34R446-R459)
> 
> ...
> 
> So if you don't want to have 2. [handled in GetStrippedOffsets code](https://github.com/loadimpact/k6/pull/1323/files#diff-4d3a948d3480dfb221e624ec54a23a34R446-R459) I can just revert [12246b4](https://github.com/loadimpact/k6/commit/12246b4449fbf22b295928a476aecf1771a10fdc) which will significantly lower the amount of code (as it will remove the whole else clause :))

It's not that I don't want it handled there, we just don't need the quadratic complexity algorithm for this use case. We can basically have no algorithm (multiplying the `Rate/TimeUnit` by the execution segment length) or a linear algorithm that returns the same results as 4). The only benefit of using the latter is that you can write unit tests that compare this fast algorithm with the results from 4.

> 4. I don't think this is a use case we have ... like what is the idea here? Either way, can we leave it for another PR later on if it's going to be done in a separate function outside of `GetStripedOffsets`?

`k6 cloud --exeecution-segment=1/2`... I'm not sure about handling this in a separate PR... The striping algorithm doesn't make a lot of sense on its own, we also need the option handling and validation, and it seems that all of that would be easier to review as a single PR?

Reply:
I am of the opinion that if you provide a not full (from 0 to 1) execution segment sequence as a cli argument you should just get an error?

There will be no point in the algorithm until we change for all the executors that will use it to ... use it which will make an even bigger PR and in particular, the arrival rate executor is being rewritten in another PR and needs this :D
And I really prefer small PRs that do one thing ;)

Reply:
This seems outdated, so resolving now